### PR TITLE
net/d_buf: remove d_buf reference from l3/l4

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/d_buf: remove d_buf reference from l3/l4

l3/l4 stack will decouple the reference of d_buf gradually, Only legacy
devices still retain d_buf support, new net devices will use d_iob

Signed-off-by: chao an <anchao@xiaomi.com>

## Impact

N/A

## Testing

iperf ipv4/ipv6 test

### CODE REVIEW

File: net/utils/net_ipchksum.c [hunk 1]
```diff
@@ -170,13 +170,11 @@ uint16_t ipv6_upperlayer_chksum(FAR struct net_driver_s *dev,
 
   /* Sum IP payload data. */
 
-#ifdef CONFIG_MM_IOB
```
Comment (xiaoxiang781216) [inline net/utils/net_ipchksum.c:173]: should we remove the check? since IPBUF also access d_iob too.

File: net/utils/net_ipchksum.c [hunk 2]
```diff
@@ -93,13 +94,11 @@ uint16_t ipv4_upperlayer_chksum(FAR struct net_driver_s *dev, uint8_t proto)
...
   if (dev->d_iob != NULL)
```
Comment (xiaoxiang781216) [inline net/utils/net_ipchksum.c:97]: should we remove the check? since IPBUF also access d_iob too.

File: net/arp/arp.h [hunk 3]
```diff
@@ -81,8 +81,10 @@
 
 /* This is a helper pointer for accessing the contents of the IP header */
 
-#define ARPBUF    ((FAR struct arp_hdr_s *)&dev->d_buf[ETH_HDRLEN])
-#define ARPIPBUF  ((FAR struct arp_iphdr_s *)&dev->d_buf[ETH_HDRLEN])
+#define ARPBUF   ((FAR struct arp_hdr_s *)\
```
Comment (xiaoxiang781216) [inline net/arp/arp.h:84]: let's move to net/arp/arp.h

File: include/nuttx/net/netdev.h [hunk 4]
```diff
@@ -161,13 +161,16 @@
  * headers
  */
 
-#define ETHBUF ((FAR struct eth_hdr_s *)&dev->d_buf[0])
+#define ETHBUF ((FAR struct eth_hdr_s *)\
```
Comment (xiaoxiang781216) [inline include/nuttx/net/netdev.h:164]: let's move to private header file inside net/?

File: net/utils/net_ipchksum.c (hunk 1 above)
Comment (anchao) [reply net/utils/net_ipchksum.c:173]: Done

File: net/utils/net_ipchksum.c (hunk 2 above)
Comment (anchao) [reply net/utils/net_ipchksum.c:97]: Done

File: net/arp/arp.h (hunk 3 above)
Comment (anchao) [reply net/arp/arp.h:84]: Done

File: include/nuttx/net/netdev.h (hunk 4 above)
Comment (anchao) [reply include/nuttx/net/netdev.h:164]: Done

File: net/arp/arp.h (hunk 3 above)
Comment (xiaoxiang781216) [reply net/arp/arp.h:86]: But no change

File: include/nuttx/net/netdev.h [hunk 5]
```diff
@@ -157,17 +157,12 @@
...
 /* There are some helper pointers for accessing the contents of the IP
  * headers
  */
 
-#define IPBUF(hl) ((FAR void *)&dev->d_buf[NET_LL_HDRLEN(dev) + (hl)])
+#define IPBUF(hl) ((FAR void *)\
+                   &dev->d_iob->io_data[CONFIG_NET_LL_GUARDSIZE + (hl)])
 
 #define IPv4BUF ((FAR struct ipv4_hdr_s *)IPBUF(0))
 #define IPv6BUF ((FAR struct ipv6_hdr_s *)IPBUF(0))
```
Comment (xiaoxiang781216) [inline include/nuttx/net/netdev.h:168]: let's move IPv4BUF/IPv6BUF/IPBUF to private header file too.

File: net/arp/arp.h (hunk 3 above)
Comment (anchao) [reply net/arp/arp.h:86]: already in net/arp/arp.h ?

File: include/nuttx/net/netdev.h (hunk 5 above)
Comment (anchao) [reply include/nuttx/net/netdev.h:168]: IPv4BUF/IPv6BUF/IPBUF are referenced by other directories

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/arm/src/lpc54xx/lpc54_ethernet.c [hunk 6]
```diff
@@ -253,6 +253,12 @@
 
 #define ETH8021QWBUF ((struct eth_8021qhdr_s *)priv->eth_dev.d_buf)
 
+/* This is a helper pointer for accessing the contents of the Ethernet
+ * header
+ */
+
+#define BUF ((FAR struct eth_hdr_s *)&dev->d_buf[0])
```
Comment (pkarashchenko) [inline arch/arm/src/lpc54xx/lpc54_ethernet.c:260]: remove `FAR`

File: arch/arm/src/tiva/lm/lm3s_ethernet.c [hunk 7]
```diff
@@ -160,6 +160,12 @@
 
 #define TIVA_MAX_MDCCLK 2500000
 
+/* This is a helper pointer for accessing the contents of the Ethernet
+ * header
+ */
+
+#define BUF ((FAR struct eth_hdr_s *)&dev->d_buf[0])
```
Comment (pkarashchenko) [inline arch/arm/src/tiva/lm/lm3s_ethernet.c:167]: remove `FAR`

File: include/nuttx/net/netdev.h [hunk 8]
```diff
@@ -157,17 +157,12 @@
...
- */
-
-#define ETHBUF ((FAR struct eth_hdr_s *)&dev->d_buf[0])
-
 /* There are some helper pointers for accessing the contents of the IP
  * headers
  */
 
-#define IPBUF(hl) ((FAR void *)&dev->d_buf[NET_LL_HDRLEN(dev) + (hl)])
+#define IPBUF(hl) ((FAR void *)\
```
Comment (pkarashchenko) [inline include/nuttx/net/netdev.h:164]: add space `) \`

File: net/arp/arp.h [hunk 9]
```diff
@@ -81,8 +83,10 @@
...
```
Comment (pkarashchenko) [inline net/arp/arp.h:86]: add space `) \`

File: net/arp/arp.h [hunk 10]
```diff
@@ -81,8 +83,10 @@
...
+                  &dev->d_iob->io_data[CONFIG_NET_LL_GUARDSIZE])
+#define ARPIPBUF ((FAR struct arp_iphdr_s *)\
```
Comment (pkarashchenko) [inline net/arp/arp.h:88]: add space `) \`

File: net/devif/devif.h [hunk 11]
```diff
@@ -233,6 +233,14 @@
 #  define DEVIF_IS_IPv6(dev) (0)
 #endif
 
+/* There are some helper pointers for accessing the contents of the Ethernet
+ * headers
+ */
+
+#define ETHBUF ((FAR struct eth_hdr_s *)\
```
Comment (pkarashchenko) [inline net/devif/devif.h:240]: add space `) \`

**Pull request merged**